### PR TITLE
Improved inventory management.  Supports:

### PR DESCRIPTION
**Pull request by mykehsd**
- on hand tracking
- on hold tracking
- minimum inventory restrictions
- minimum/maximum purchase requirements
- will be easly converted to store level customizations when that's supported in sylius

### CODE REVIEW

**Comment by mykehsd** [comment]
This wasn't 100% yet but since I saw #2354 I figured I would post it

File: src/Sylius/Bundle/CoreBundle/Checkout/Step/InventoryStep.php [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+use Sylius\Component\Inventory\Manager\InsufficientRequirementsException;
+
+/**
+ * Inventory step.
+ *
+ * Checks to make sure all inventory is avilable
+ *
+ * @author Myke Hines <myke@webhines.com>
+ */
+class InventoryStep extends CheckoutStep
```
Comment (stloyd) [inline src/Sylius/Bundle/CoreBundle/Checkout/Step/InventoryStep.php]: I don't think this is good idea. IIRC we use CartResolver to check such things. I.e. have a look at https://github.com/Sylius/Sylius/pull/1935 that was one of ideas how to improve this stuff.

File: src/Sylius/Bundle/CoreBundle/Resources/config/validation.xml [hunk 2]
```diff
@@ -71,19 +71,57 @@
...
+                <option name="groups">sylius</option>
+            </constraint>
+        </property>        
+        <property name="maxQuantityInCart">
+            <constraint name="NotBlank">
+                <option name="message">sylius.stock.maxQuantityInCart.not_blank</option>
+                <option name="groups">sylius</option>
+            </constraint>
+            <constraint name="Range">
+                <option name="min">0</option>
```
Comment (stloyd) [inline src/Sylius/Bundle/CoreBundle/Resources/config/validation.xml]: You need to validate that `max` is not smaller than `min` ;)

File: src/Sylius/Bundle/FixturesBundle/DataFixtures/ORM/LoadProductsData.php [hunk 3]
```diff
@@ -280,6 +284,16 @@ protected function addMasterVariant(ProductInterface $product, $sku = null)
         );
         $variant->addImage($image);
 
+        $stock = new ProductVariantStock();
```
Comment (stloyd) [inline src/Sylius/Bundle/FixturesBundle/DataFixtures/ORM/LoadProductsData.php]: Calling model directly sounds wrong.

File: src/Sylius/Bundle/InventoryBundle/Form/Type/StockType.php [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    /**
+     * {@inheritdoc}
+     */
+    public function buildForm(FormBuilderInterface $builder, array $options)
+    {
+        $builder
+            ->add('manage_stock', 'checkbox')
+            ->add('onHand')
+            ->add('allowBackorders', 'checkbox')
+            ->add('minQuantityInCart')
```
Comment (stloyd) [inline src/Sylius/Bundle/InventoryBundle/Form/Type/StockType.php]: Here & below you should add `number` or `integer` as type.

File: src/Sylius/Bundle/InventoryBundle/Resources/config/doctrine/model/Stock.orm.xml [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+                                      http://doctrine-project.org/schemas/orm/doctrine-mapping.xsd">
+
+    <mapped-superclass name="Sylius\Component\Inventory\Model\Stock" table="sylius_inventory_stock">
+        <id name="id" column="id" type="integer">
+            <generator strategy="AUTO" />
+        </id>
+      
+        <field name="manageStock" column="manage_stock" type="boolean" nullable="false" />
+        <field name="onHand" column="on_hand" type="float" />
+        <field name="onHold" column="on_hold1" type="integer" />
```
Comment (stloyd) [inline src/Sylius/Bundle/InventoryBundle/Resources/config/doctrine/model/Stock.orm.xml]: Wrong rebase?

File: src/Sylius/Bundle/InventoryBundle/Resources/config/services.xml [hunk 6]
```diff
@@ -18,25 +18,27 @@
...
+        <parameter key="sylius.manager.inventory.default.class">Sylius\Component\Inventory\Manager\InventoryManager</parameter>
         <parameter key="sylius.inventory_operator.noop.class">Sylius\Component\Inventory\Operator\NoopInventoryOperator</parameter>
         <parameter key="sylius.inventory_operator.default.class">Sylius\Component\Inventory\Operator\InventoryOperator</parameter>
         <parameter key="sylius.backorders_handler.class">Sylius\Component\Inventory\Operator\BackordersHandler</parameter>
         <parameter key="sylius.inventory_unit_factory.class">Sylius\Component\Inventory\Factory\InventoryUnitFactory</parameter>
         <parameter key="sylius.listener.inventory.class">Sylius\Bundle\InventoryBundle\EventListener\InventoryChangeListener</parameter>
         <parameter key="sylius.validator.in_stock.class">Sylius\Bundle\InventoryBundle\Validator\Constraints\InStockValidator</parameter>
+        <parameter key="sylius.form.type.inventory_stock.class">Sylius\Bundle\InventoryBundle\Form\Type\StockType</parameter>
+        <parameter key="sylius.model.inventory_stock.class">Sylius\Component\Inventory\Model\Stock</parameter>
```
Comment (stloyd) [inline src/Sylius/Bundle/InventoryBundle/Resources/config/services.xml]: This should be in configuration not here.

File: src/Sylius/Component/Core/Model/ProductVariantStockInterface.php [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sylius\Component\Core\Model;
+
+use Doctrine\Common\Collections\Collection;
+use Sylius\Component\Product\Model\VariantInterface as BaseVariantInterface;
```
Comment (stloyd) [inline src/Sylius/Component/Core/Model/ProductVariantStockInterface.php]: This & above line is not used.

**Comment by mykehsd** [comment]
@stloyd Thanks for your input!

File: src/Sylius/Bundle/CoreBundle/Checkout/Step/InventoryStep.php [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+
+use Sylius\Bundle\FlowBundle\Process\Context\ProcessContextInterface;
+use Symfony\Component\HttpFoundation\RedirectResponse;
+use Sylius\Component\Core\SyliusCheckoutEvents;
+use Sylius\Component\Inventory\Manager\InsufficientRequirementsException;
+
+/**
+ * Inventory step.
+ *
+ * Checks to make sure all inventory is avilable
```
Comment (pborreli) [inline src/Sylius/Bundle/CoreBundle/Checkout/Step/InventoryStep.php]: avilable -> available

**Comment by mykehsd** [comment]
@pjedrzejewski let me know if I should invest any effort to fix a few behat issues.

**Comment by stloyd** [comment]
I don't like usage of checkout step here, IMO it should be done in listener or something like we already have in resolver...

@pjedrzejewski @aRn0D feedback?

**Comment by mykehsd** [comment]
@pjedrzejewski ?

**Comment by pjedrzejewski** [comment]
I like this a lot! I'd love to get some of these improvements combined with our multi-stock. Hopefully it will get extracted soon. Then we should look at this. Thanks for your work Myke! :+1:

**Comment by mykehsd** [comment]
@pjedrzejewski Let me know when your multi stock branch is out and I'll help integrate

**Comment by loevstroem** [comment]
@mykehsd It's already available from #2354 I'm focusing on multi-channel, and will pickup multi-stock afterwards if you don't

**Comment by pjedrzejewski** [comment]
@loevstroem Awesome, multi-stock should come after multi-channel. Kristian, good idea would be to just extrat what we did and iterate on it together, I have some time and ideas now to help you with pushing this forward. :)

**Pull request closed without merging**